### PR TITLE
[Agent Management] Add Remote Config Validation

### PR DESCRIPTION
**Pull request by spartan0x117**
#### PR Description
Refactors `agentmanagement.go` so `getRemoteConfig` accepts `remoteConfigProvider`s, which allows for more testing. Adds tests for some cases.

Also splits out the code that parses the flagset (the second time) and sets integrations version in `pkg/config/config.go`, specifically `load`, into a separate function which `getRemoteConfig` also calls. `getRemoteConfig` is also modified to take the parameters needed by this function (the `flagSet`, `args`, and the config `path`).

This is done in order to allow for semantic validation of the remote config inside the agent management code. Otherwise, if the API responds with a semantically invalid config (i.e. it's valid yaml, but it's not a valid agent management config) the _invalid_ remote config will be cached and any agents attempting to use it upon startup will panic.

Finally, this adds a new metric `agent_remote_config_invalid_total`, which tracks the number of invalid configs that were returned by the API, with possible labels `invalid_yaml` and `invalid_agent_config`.

#### PR Checklist

- [ ] CHANGELOG updated
- [ ] Documentation added
- [X] Tests updated

### CODE REVIEW

File: pkg/config/agentmanagement.go [hunk 1]
```diff
@@ -116,6 +91,78 @@ func fetchConfig(c *Config) ([]byte, error) {
...
+	err = LoadBytes(remoteConfigBytes, expandEnvVars, &remoteConfig)
+	if err != nil {
+		level.Error(log).Log("msg", "could not load the response from the API, falling back to cache", "err", err)
+		instrumentation.InstrumentInvalidRemoteConfig("invalid_yaml")
+		return configProvider.GetCachedRemoteConfig(expandEnvVars)
+	}
+
+	// this is done in order to validate the config semantically
+	if err = applyIntegrationValuesFromFlagset(fs, args, configPath, &remoteConfig); err != nil {
+		level.Error(log).Log("msg", "could not apply integration version to config", "err", err)
```
Comment (jcreixell) [inline pkg/config/agentmanagement.go:137]: probably instrument here too with reason "invalid_flags" (arguably this should never happen but who knows)

File: pkg/config/agentmanagement.go [hunk 2]
```diff
@@ -116,6 +91,78 @@ func fetchConfig(c *Config) ([]byte, error) {
...
+
+	err = LoadBytes(remoteConfigBytes, expandEnvVars, &remoteConfig)
+	if err != nil {
+		level.Error(log).Log("msg", "could not load the response from the API, falling back to cache", "err", err)
+		instrumentation.InstrumentInvalidRemoteConfig("invalid_yaml")
+		return configProvider.GetCachedRemoteConfig(expandEnvVars)
+	}
+
+	// this is done in order to validate the config semantically
+	if err = applyIntegrationValuesFromFlagset(fs, args, configPath, &remoteConfig); err != nil {
```
Comment (jcreixell) [inline pkg/config/agentmanagement.go:136]: for a follow up refactoring PR: some of this logic is so entangled with config.go that it should probably moved there (these two files know and depend on each other, which isn't great from the modularity POV)

File: pkg/config/agentmanagement.go [hunk 3]
```diff
@@ -116,6 +91,78 @@ func fetchConfig(c *Config) ([]byte, error) {
...
 }
 
+type labelMap map[string]string
+
+type RemoteConfiguration struct {
+	Labels    labelMap `yaml:"labels"`
+	Namespace string   `yaml:"namespace"`
+}
+
+type AgentManagement struct {
```
Comment (jcreixell) [inline pkg/config/agentmanagement.go:101]: consider renaming this to AgentManagementConfig``

**Review by jcreixell — COMMENTED** [review]
LGTM, I am gonna do a few tests locally and if everything looks good I will give it a :+1:

**Review by jcreixell — APPROVED** [review]
Good job, I have tested this thoroughly and seems to work as expected.

I found an unrelated issue that we will need to fix separately (adding, removing and adding again log scrape configs seems to cause an issue with registering prometheus metrics in promtail, I will create a separate issue for that)

File: pkg/config/agentmanagement_test.go [hunk 4]
```diff
@@ -87,27 +124,14 @@ func TestMissingCacheLocation(t *testing.T) {
 	assert.Error(t, invalidConfig.Validate())
 }
 
-func TestGetCachedRemoteConfig(t *testing.T) {
-	cwd := filepath.Clean("./testdata/")
-	_, err := getCachedRemoteConfig(cwd, false)
-	assert.NoError(t, err)
-}
+// func TestGetCachedRemoteConfig(t *testing.T) {
```
Comment (marctc) [inline pkg/config/agentmanagement_test.go:127]: We could probably delete this code

**Comment by jcreixell** [comment]
DNM, adding some changes

**Comment by jcreixell** [comment]
I am done with my changes, @marctc feel free to merge if you are happy with this

**Review by marctc — APPROVED** [review]

**Pull request merged**